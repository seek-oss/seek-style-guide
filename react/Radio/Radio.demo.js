import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Radio from './Radio';
import Text from '../Text/Text';
import * as sketch from './Radio.sketch';
import demoStyles from './Radio.demo.less';

class RadioContainer extends Component {
  static propTypes = {
    component: PropTypes.func.isRequired,
    componentProps: PropTypes.object.isRequired
  };

  constructor() {
    super();

    this.state = {
      value: ''
    };
  }

  handleYes = () => {
    this.setState({
      value: 'Yes'
    });
  };

  handleNo = () => {
    this.setState({
      value: 'No'
    });
  };

  handleMaybe = () => {
    this.setState({
      value: 'Maybe'
    });
  };

  render() {
    const { component: DemoComponent, componentProps } = this.props;
    const { value } = this.state;

    return (
      <div className={demoStyles.root}>
        <Text subheading>Still in role</Text>
        <DemoComponent
          {...componentProps}
          id="Yes"
          label="Yes"
          checked={value === 'Yes'}
          onChange={this.handleYes}
        />
        <DemoComponent
          {...componentProps}
          id="No"
          label="No"
          checked={value === 'No'}
          onChange={this.handleNo}
        />
        <DemoComponent
          {...componentProps}
          id="Maybe"
          label="Maybe - Very long label which goes over many lines."
          checked={value === 'Maybe'}
          onChange={this.handleMaybe}
        />
      </div>
    );
  }
}

export default {
  route: '/radio',
  title: 'Radio',
  category: 'Form',
  component: Radio,
  container: RadioContainer,
  sketch,
  initialProps: {
    name: 'stillInRole',
    // Documentation only:
    id: 'myRadio',
    checked: false,
    onChange: () => {}
  },
  options: []
};
