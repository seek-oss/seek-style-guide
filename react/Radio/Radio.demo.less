.root {
  display: flex;
  flex-direction: column;
  width: 175px;

  > :global(*) {
    min-width: 150px; // :(
  }
}
