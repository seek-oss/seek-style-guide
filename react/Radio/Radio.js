import styles from './Radio.less';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classnames from 'classnames';
import Text from '../Text/Text';

function combineClassNames(props = {}, ...classNames) {
  const { className, ...restProps } = props;

  return {
    className: classnames.apply(null, [...classNames, className]), // eslint-disable-line no-useless-call
    ...restProps
  };
}

export default class Radio extends Component {
  static displayName = 'Radio';

  static propTypes = {
    id: PropTypes.string.isRequired,
    value: PropTypes.string,
    checked: PropTypes.bool.isRequired,
    onChange: PropTypes.func.isRequired,
    onFocus: PropTypes.func,
    onBlur: PropTypes.func,
    className: PropTypes.string,
    label: PropTypes.string,
    labelProps: PropTypes.object,
    inputProps: PropTypes.object
  };

  static defaultProps = {
    className: '',
    label: '',
    inputProps: {}
  };

  renderInput() {
    const {
      id,
      value,
      checked,
      onChange,
      onFocus,
      onBlur,
      inputProps
    } = this.props;
    const allInputProps = {
      id,
      value,
      checked,
      onChange,
      onFocus,
      onBlur,
      ...combineClassNames(inputProps, styles.input),
      type: 'radio'
    };

    return <input {...allInputProps} />;
  }

  render() {
    const { label, labelProps, id, className } = this.props;
    const allLabelProps = {
      htmlFor: id,
      ...combineClassNames(labelProps, styles.label)
    };

    return (
      <div className={className}>
        <label {...allLabelProps}>
          {this.renderInput()}
          <svg
            className={styles.svg}
            viewBox="0 0 200 200"
            xmlns="http://www.w3.org/2000/svg"
            focusable="false"
          >
            <circle
              className={classnames(styles.circle, styles.circle_isHover)}
              cx="100"
              cy="100"
              r="100"
            />
            <circle
              className={classnames(styles.circle, styles.circle_isSelected)}
              cx="100"
              cy="100"
              r="100"
            />
          </svg>
          <Text raw={true} baseline={false} className={styles.labelText}>
            {label}
          </Text>
        </label>
      </div>
    );
  }
}
