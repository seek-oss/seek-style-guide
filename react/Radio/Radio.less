@import (reference) "~seek-style-guide/theme";
@circumference: @row-height * 5;
@touchableMargin: (@touchableTextHeight - @circumference) / 2;

.label {
  display: flex;
  cursor: pointer;
  min-height: @touchableTextHeight;

  :global(*) {
    pointer-events: none;
  }
}

.input {
  opacity: 0;
  position: absolute;
  width: @circumference;
  height: @circumference;
  border: none;
  margin: @touchableMargin 0 0 0;
}

.labelText {
  @media @desktop {
    padding: (
        ((@interaction-type-row-span - @standard-type-row-span) / 2) *
          @row-height
      )
      0;
  }
  @media @mobile {
    padding: (
        ((@interaction-type-row-span - @standard-type-row-span-mobile) / 2) *
          @row-height
      )
      0;
  }

  .input:checked + .svg & {
    font-weight: @sk-bold;
  }
}

.svg {
  flex-shrink: 0;
  background-color: @sk-white;
  width: @circumference;
  height: @circumference;
  border: @field-border-width solid @sk-mid-gray-light;
  border-radius: 100%;
  padding: 5px;
  margin-top: @touchableMargin;
  margin-right: @gutter-width / 2;

  .input:not([readonly]):not([disabled]):focus + .svg & {
    .focus();
    border-radius: 100%;
  }
}

.circle {
  transition: opacity 0.2s ease, @spring-transition;
  transform-origin: 50% 50%;
  opacity: 0;

  .input:active + .svg & {
    transform: scale(0.9);
  }

  &.circle_isSelected {
    fill: @sk-form-accent;

    .input:checked + .svg & {
      opacity: 1;
    }
  }

  &.circle_isHover {
    fill: @sk-gray-light;

    .input:focus:not(:checked) + .svg &,
    .input:hover:not(:checked) + .svg & {
      opacity: 1;
    }
  }
}
