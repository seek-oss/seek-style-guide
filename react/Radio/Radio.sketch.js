import React from 'react';
import PropTypes from 'prop-types';
import Radio from './Radio';
import styles from './Radio.demo.less';

const Container = ({ children }) => (
  <div className={styles.root}>{children}</div>
);
Container.propTypes = {
  children: PropTypes.node.isRequired
};

const noop = () => {};

export const symbols = {
  'Radio/Checked': (
    <Container>
      <Radio id="radio2" label="Radio" checked={true} onChange={noop} />
    </Container>
  ),
  'Radio/Unchecked': (
    <Container>
      <Radio id="radio1" label="Radio" checked={false} onChange={noop} />
    </Container>
  )
};
