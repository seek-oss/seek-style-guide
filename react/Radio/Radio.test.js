import React from 'react';
import { shallow } from 'enzyme';
import Radio from './Radio';

describe('Radio', () => {
  const requiredProps = {
    id: 'testRadio',
    label: 'Still in role',
    checked: false,
    onChange: () => {}
  };

  it('should render with simple props', () => {
    const wrapper = shallow(<Radio {...requiredProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with className', () => {
    const className = 'testClassname';
    const wrapper = shallow(<Radio {...requiredProps} className={className} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render as checked', () => {
    const wrapper = shallow(<Radio {...requiredProps} checked={true} />);
    const inputChecked = wrapper.find('input').prop('checked');
    expect(inputChecked).toEqual(true);
  });

  it('should pass through the value', () => {
    const wrapper = shallow(
      <Radio {...requiredProps} value="foo" checked={true} />
    );
    const inputValue = wrapper.find('input').prop('value');
    expect(inputValue).toEqual('foo');
  });

  it('should invoke the change handler', () => {
    const onChange = jest.fn();
    const props = { ...requiredProps, onChange, checked: false };
    const checkedEvent = { target: { checked: true } };
    const wrapper = shallow(<Radio {...props} />);
    wrapper.find('input').simulate('change', checkedEvent);
    expect(onChange).toBeCalledWith(checkedEvent);
  });

  it('should invoke the focus handler', () => {
    const onFocus = jest.fn();
    const props = { ...requiredProps, onFocus };
    const wrapper = shallow(<Radio {...props} />);
    wrapper.find('input').simulate('focus');
    expect(onFocus.mock.calls.length).toBe(1);
  });

  it('should invoke the blur handler', () => {
    const onBlur = jest.fn();
    const props = { ...requiredProps, onBlur };
    const wrapper = shallow(<Radio {...props} />);
    wrapper.find('input').simulate('blur');
    expect(onBlur.mock.calls.length).toBe(1);
  });

  describe('inputProps', () => {
    it('should invoke the change handler', () => {
      const onChange = jest.fn();
      const props = {
        ...requiredProps,
        inputProps: { onChange, checked: false }
      };
      const checkedEvent = { target: { checked: true } };
      const wrapper = shallow(<Radio {...props} />);
      wrapper.find('input').simulate('change', checkedEvent);
      expect(onChange).toBeCalledWith(checkedEvent);
    });

    it('should invoke the focus handler', () => {
      const onFocus = jest.fn();
      const props = { ...requiredProps, inputProps: { onFocus } };
      const wrapper = shallow(<Radio {...props} />);
      wrapper.find('input').simulate('focus');
      expect(onFocus.mock.calls.length).toBe(1);
    });

    it('should invoke the blur handler', () => {
      const onBlur = jest.fn();
      const props = { ...requiredProps, inputProps: { onBlur } };
      const wrapper = shallow(<Radio {...props} />);
      wrapper.find('input').simulate('blur');
      expect(onBlur.mock.calls.length).toBe(1);
    });

    it('should pass through other props to the input', () => {
      const wrapper = shallow(
        <Radio
          {...requiredProps}
          inputProps={{ checked: true, 'data-automation': 'first-name-field' }}
        />
      );
      expect(wrapper).toMatchSnapshot();
    });
  });
});
