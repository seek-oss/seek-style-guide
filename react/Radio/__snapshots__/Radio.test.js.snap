// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio inputProps should pass through other props to the input 1`] = `
<div
  className=""
>
  <label
    className="Radio__label"
    htmlFor="testRadio"
  >
    <input
      checked={true}
      className="Radio__input"
      data-automation="first-name-field"
      id="testRadio"
      onChange={[Function]}
      type="radio"
    />
    <svg
      className="Radio__svg"
      focusable="false"
      viewBox="0 0 200 200"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="Radio__circle Radio__circle_isHover"
        cx="100"
        cy="100"
        r="100"
      />
      <circle
        className="Radio__circle Radio__circle_isSelected"
        cx="100"
        cy="100"
        r="100"
      />
    </svg>
    <Text
      baseline={false}
      className="Radio__labelText"
      raw={true}
    >
      Still in role
    </Text>
  </label>
</div>
`;

exports[`Radio should render with className 1`] = `
<div
  className="testClassname"
>
  <label
    className="Radio__label"
    htmlFor="testRadio"
  >
    <input
      checked={false}
      className="Radio__input"
      id="testRadio"
      onChange={[Function]}
      type="radio"
    />
    <svg
      className="Radio__svg"
      focusable="false"
      viewBox="0 0 200 200"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="Radio__circle Radio__circle_isHover"
        cx="100"
        cy="100"
        r="100"
      />
      <circle
        className="Radio__circle Radio__circle_isSelected"
        cx="100"
        cy="100"
        r="100"
      />
    </svg>
    <Text
      baseline={false}
      className="Radio__labelText"
      raw={true}
    >
      Still in role
    </Text>
  </label>
</div>
`;

exports[`Radio should render with simple props 1`] = `
<div
  className=""
>
  <label
    className="Radio__label"
    htmlFor="testRadio"
  >
    <input
      checked={false}
      className="Radio__input"
      id="testRadio"
      onChange={[Function]}
      type="radio"
    />
    <svg
      className="Radio__svg"
      focusable="false"
      viewBox="0 0 200 200"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="Radio__circle Radio__circle_isHover"
        cx="100"
        cy="100"
        r="100"
      />
      <circle
        className="Radio__circle Radio__circle_isSelected"
        cx="100"
        cy="100"
        r="100"
      />
    </svg>
    <Text
      baseline={false}
      className="Radio__labelText"
      raw={true}
    >
      Still in role
    </Text>
  </label>
</div>
`;
